### ISSUE TITLE
Feature request: fuzzing absolute URLs (separate `<url>` from what is actually connected to)

### ISSUE THREAD

**Issue by Prinzhorn** [xmendez/wfuzz#121]
# Issue template

## Context

**Please check:**

- [x] I've read the docs for [Wfuzz](http://wfuzz.readthedocs.io/)

**Please describe your local environment:**

Wfuzz version: 2.3.2
Python version: Python 3.6.7
OS: Ubuntu 18.04.1

## Report

**What is the current behavior?**

Using `wfuzz [...] http://hostname/PATH` always results in wfuzz connecting to `hostname` and issuing a `GET /PATH` request (or whatever method you chose).

**What is the expected or desired behavior?**

TLDR: I want to use the WFUZZ keyword in this place: `GET WFUZZ HTTP/1.1` and in the `<url>` argument at the same time.

I would love to be able to fuzz (misconfigured) endpoints that support absolute URLs. That means we would need a new option to specify the host and port to connect to - completely separate from the URL that is used in the request.

Essentially I want to make requests to `actual_hostname` that look like this `GET http://different_hostname/PATH`. I'm aware that you can already fuzz the Host header, but that's not what I'm asking for.

Burp intruder only supports half of this. You can specify the server to connect to but you cannot fuzz it.

![selection_445](https://user-images.githubusercontent.com/679144/52205350-13357300-2877-11e9-84be-a88befe41f91.png)

So with Burp I can connect to any endpoint and then send completely unrelated HTTP requests. I need to get one step further and also fuzz the endpoint that the TCP connection is made to.

**Comment by xmendez** [comment]
Hi @Prinzhorn ,

Are you trying to fuzz the virtualhosts of a server?

**Comment by Prinzhorn** [comment]
> Are you trying to fuzz the virtualhosts of a server?

No, these are completely unrelated servers (in-scope for certain bug bounty programs) that I want to fuzz for a particular reverse proxy misconfiguration. Here's a related post https://portswigger.net/blog/cracking-the-lens-targeting-https-hidden-attack-surface (they're using ZGrab)

So I need to make a handful of similar requests to many different servers. WFuzz might be the wrong tool for the job and if you think it doesn't fit the scope and it won't be useful for others feel free to close this issue. I'm not sure if the Python network stack would even easily allow making these requests?

I'm now rolling my own script that spawns `ncat` processes and pipes raw requests from a template file.

**Comment by xmendez** [comment]
Ok, I got you now. I will take a look later.

**Comment by Prinzhorn** [comment]
Thank you! But maybe we should wait and see if others find it useful as well. There are likely better thinks you can put your time into.

I've since rolled my own setup and realized that I would've likely hit some other WFuzz limitations anyway. For example some of my fuzzing lists need to be customized based on the target itself. E.g. parts of the list of `X-Forwarded-Host` headers depends on the target that the connection is made to. My script is able to create all permutations I need while being able to fully customize the raw HTTP request.

**Comment by xmendez** [comment]
Actually, pycurl supports connecting to a specific IP. See https://curl.haxx.se/libcurl/c/CURLOPT_CONNECT_TO.html
So, it's quite simple to support this.

I am always interested in supporting advanced uses cases in wfuzz. So, please keep sharing ideas. Do you have your script in a github project?

**Comment by Prinzhorn** [comment]
That sounds great.

> Do you have your script in a github project?

No, since it is just a hacky script right now. What it does is create raw HTTP requests from a mustache template like this

```
{{address}} {{port}} {{sslEnabled}}
GET {{url}} HTTP/1.1
{{#headers}}{{#value}}{{name}}: {{value}}
{{/value}}{{/headers}}
Cache-Control: no-transform
Connection: close


```

which are then fed into a worker that executes them in parallel (that's what the first line of the template is for).

And the permutations of different `{{url}}` and `{{headers}}` is what wfuzz might be capable of. But like I said certain values in the list depend on others. E.g. I have a list of hostnames for each ip which are used to build `{{url}}` and some of the headers in some of the iterations. So basically the fuzz list is a template as well or at least each line can have placeholders.

**Comment by xmendez** [comment]
Please, check if this issue has been addressed with Wfuzz 2.4.

**Comment by Prinzhorn** [comment]
It works as advertised, thanks for that!

One thing would be needed to fully do what this issue was originally asking for, namely

```
--ip FUZZ:FUZ2Z
```

For example when you have a list of reverse proxies you want to probe for a list of hostnames.

**Comment by xmendez** [comment]
Yeah, that makes sense, leaving the issue open
